Sources/Target/Windows/Thread.cpp

```diff
@@ -81,14 +81,54 @@ void Thread::updateState(DEBUG_EVENT const &de) {
   switch (de.dwDebugEventCode) {
   case EXCEPTION_DEBUG_EVENT:
     _state = kStopped;
-    // We always specify kReasonBreakpoint here, but we should instead analyze
-    // what the ExceptionCode is, and set the stop reason accordingly.
-    _stopInfo.event = StopInfo::kEventStop;
-    _stopInfo.reason = StopInfo::kReasonBreakpoint;
+
     DS2LOG(
         Debug, "exception from inferior, code=%s, address=%p",
         Stringify::ExceptionCode(de.u.Exception.ExceptionRecord.ExceptionCode),
         de.u.Exception.ExceptionRecord.ExceptionAddress);
+
+    switch (de.u.Exception.ExceptionRecord.ExceptionCode) {
+    case EXCEPTION_BREAKPOINT:
+    case EXCEPTION_SINGLE_STEP:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonBreakpoint;
+      break;
+
+    case EXCEPTION_ACCESS_VIOLATION:
+    case EXCEPTION_ARRAY_BOUNDS_EXCEEDED:
+    case EXCEPTION_IN_PAGE_ERROR:
+    case EXCEPTION_STACK_OVERFLOW:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonMemoryError;
+      break;
+
+    case EXCEPTION_FLT_DENORMAL_OPERAND:
+    case EXCEPTION_FLT_DIVIDE_BY_ZERO:
+    case EXCEPTION_FLT_INEXACT_RESULT:
+    case EXCEPTION_FLT_INVALID_OPERATION:
+    case EXCEPTION_FLT_OVERFLOW:
+    case EXCEPTION_FLT_STACK_CHECK:
+    case EXCEPTION_FLT_UNDERFLOW:
+    case EXCEPTION_INT_DIVIDE_BY_ZERO:
+    case EXCEPTION_INT_OVERFLOW:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonMathError;
+      break;
+
+    case EXCEPTION_ILLEGAL_INSTRUCTION:
+    case EXCEPTION_PRIV_INSTRUCTION:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonInstructionError;
+      break;
+
+    case EXCEPTION_INVALID_DISPOSITION:
+    case EXCEPTION_NONCONTINUABLE_EXCEPTION:
+    default:
+      DS2BUG("unsupported exception code: %s",
+             Stringify::ExceptionCode(
+                 de.u.Exception.ExceptionRecord.ExceptionCode));
```

Comment:
This'll probably happen a *lot*. You should probably communicate the error instead of crashing.